src/v/cluster/rm_stm.cc

```diff
@@ -783,6 +792,108 @@ ss::future<bool> rm_stm::sync(model::timeout_clock::duration timeout) {
     co_return ready;
 }
 
+void rm_stm::track_tx(
+  model::producer_identity pid, int32_t transaction_timeout_ms) {
+    _mem_state.livetime[pid] = livetime{
+      .timeout = std::chrono::milliseconds(transaction_timeout_ms),
+      .last_update = clock_type::now()};
+    auto deadline = _mem_state.livetime[pid].deadline();
+    if (
+      (auto_abort_timer.armed() && auto_abort_timer.get_timeout() > deadline)
+      || !auto_abort_timer.armed()) {
+        auto_abort_timer.cancel();
+        auto_abort_timer.arm(deadline);
+    }
+}
+
+void rm_stm::abort_old_txes() {
+    std::vector<model::producer_identity> pids;
+    for (auto& [k, _] : _mem_state.estimated) {
+        pids.push_back(k);
+    }
+    for (auto& [k, _] : _mem_state.tx_start) {
+        pids.push_back(k);
+    }
+    for (auto& [k, _] : _log_state.ongoing_map) {
+        pids.push_back(k);
+    }
+    std::vector<model::producer_identity> expired;
+    for (auto pid : pids) {
+        if (_log_state.prepared.contains(pid)) {
+            continue;
+        }
+        if (_mem_state.preparing.contains(pid)) {
+            continue;
+        }
+        if (_mem_state.expected.contains(pid)) {
+            auto livetime_it = _mem_state.livetime.find(pid);
+            if (livetime_it != _mem_state.livetime.end()) {
+                if (livetime_it->second.deadline() > clock_type::now()) {
+                    continue;
+                }
+            }
+        }
+        expired.push_back(pid);
+    }
+    (void)abort_old_txes(expired);
+}
+
+ss::future<>
+rm_stm::abort_old_txes(std::vector<model::producer_identity> pids) {
```

Comment:
I think it's a rare situation but even if it happens nothing can go wrong because the purpose of `abort_old_txes` is to collect the candidates (possible expired transactions) then for each pid we run `try_abort_old_tx` which takes the pid's lock and double checks that it's still idle. So if there are multiple `abort_old_txes` they will be "synchronised" on that lock and nothing wrong will happen. Don't think it's worth to introduce additional semaphore.

Reply:
got it. that sounds reasonable. how rare do you think it is? it seems related to the time it takes to replicate (effectively a debounce), GCing items out of  _mem_state.tx_start etc..., and the average deadline. I don't have a good sense for that relationship

Reply:
rare :) it may happen when a client pass transaction_timeout_ms which is less that number of timed out transactions at that time x time of (2x flushes + cluster wide RTT). Default value for transaction_timeout_ms is a minute and flushes + RTT are on the ms scale